File: compile/Compiler.java
package compile;

import file.Script;
import tools.ToolBox;

import java.util.ArrayList;

/**
 * Created by John Elizarraras on 2/28/2016.
 */
public class Compiler {

    ArrayList<CompileError> errors = new ArrayList<>();
    Script file;
    String path;
    int aliasNum;
    /**
     * Creates a new Compiler with a path of main.txt
     */
    public Compiler() {
        path = "main.txt";
        file = new Script(path);
        compile();
        new CompilerViewer(this, "compileErrors");
        aliasNum = 0;
    }

    /**
     * Creates a new Compiler
     *
     * @param path the path of the file
     */
    public Compiler(String path) {
        this.path = path;
        file = new Script(path);
        compile();
        new CompilerViewer(this, "compileErrors");
        aliasNum = 0;
    }

    public Compiler(String path, String fileName) {
        this.path = path;
        file = new Script(path);
        compile();
        new CompilerViewer(this, fileName);
        aliasNum = 0;
    }

    public String getString(){
        return path;
    }

    /**
     * adds an error
     *
     * @param line         the line of the error
     * @param errorMessage the error message
     */
    public void addError(int line, String errorMessage) {
        errors.add(new CompileError(line, errorMessage));
    }
    /**
     * compiles all of the file
     */
    private void compile() {
        for (int i = 0; i < file.length(); i++) {
            String line = file.fullLine();
            if (isCommand(file.command())) {
                switch (file.command().toLowerCase()) {
                    case "run":
                        checkRun(i);
                        break;
                    case "loop":
                        checkLoop(i);
                        break;
                    case "display":
                        checkDisplay(i);
                        break;
                    case "{":
                        checkOpenBracket(i);
                        break;
                    case "}":
                        checkCloseBracket(i);
                        break;
                    case "alias":
                        checkAlias(i);
                        break;
                    default:
                        addError(i, "Not a command");
                        break;
                }
            } else {
                addError(i, "This is not a command");
            }
            file.nextLine();
        }
    }

    private void checkAlias(int index) {
        String[] vars = file.object(index);

        if (!ToolBox.isFile(ToolBox.combine(vars, true, false))) {
            addError(index, "String given is not a file!");
        }
        else{
            Compiler comp = new Compiler(ToolBox.combine(vars, true, false), "compileErrors" + aliasNum);
            if(comp.errors() > 0) addError(index, "Alias has an error!");
            aliasNum++;
        }
    }

    /**
     * checks if the string is a command
     *
     * @param posCommand the possible command
     * @return true if the string is an accepted command
     */
    private boolean isCommand(String posCommand) {
        boolean result = true;
        switch (posCommand.toLowerCase()) {
            case "loop":
                break;
            case "run":
                break;
            case "display":
                break;
            case "{":
                break;
            case "}":
                break;
            case "alias":
                break;
            default:
                result = false;
                break;
        }
        return result;
    }

    /**
     * checks if the run command is in the right format
     *
     * @param index the index of the command
     */
    public void checkRun(int index) {

    }

    /**
     * checks if the loop is in the correct format
     *
     * @param index the index of the loop command
     */
    public void checkLoop(int index) {
        String[] vars = file.object(index);
        if (vars.length < 2) addError(index, "Not enough modifiers");
        else if (vars.length > 2) addError(index, "Too many modifiers");
        else {
            if (isNumber(vars[0].substring(0, vars[0].length() - 1))) {
            } else {
                addError(index, "Time is not a number");
            }
            switch (vars[0].substring(vars[0].length() - 1)) {
                case "s":
                    break;
                case "m":
                    break;
                case "h":
                    break;
                case "d":
                    break;
                default:
                    addError(index, "Not an acceptable time format.");
                    break;
            }

            if (!isNumber(vars[1])) addError(index, "Second modifier is not a number");
            else {
                if (Integer.parseInt(vars[1]) < 0) addError(index, "Second modifier is negative");
            }
            try {
                if (!file.fullLine(index + 1).equals("{")) addError(index + 1, "Is not a '{'");

                boolean needsError = true;
                for (int i = index + 2; i < file.length(); i++) {
                    if (file.fullLine(i).equals("{")) {
                        addError(i, "Another '{' before a '}'");
                    } else if (file.fullLine(i).equals("}")) {
                        needsError = false;
                        break;
                    }
                }
                if (needsError) addError(file.length(), "No closing '}'");
            } catch (IndexOutOfBoundsException e) {
                addError(file.length(), "Loop doesn't open/close");
            }
        }
    }

    /**
     * checks if the display command is in the right format
     *
     * @param index the index of the run command
     */
    private void checkDisplay(int index) {

    }

    private void checkOpenBracket(int index) {
        try {
            if (!lineContains("loop", file.fullLine(index - 1), 0, 4))
                addError(index, "Open bracket without loop before");
        } catch (Exception e) {
            addError(index, "Open bracket without loop before");
        }
    }

    private void checkCloseBracket(int index) {
        try {
            for (int i = index - 1; i >= 0; i--) {
                if (lineContains("loop", file.fullLine(i), 0, 4)) return;
                else if (lineContains("}", file.fullLine(i), 0, 1)) addError(i, "'}' before '{'");
            }
            addError(index, "'}' with no opening bracket");
        } catch (Exception e) {
            addError(index, "Closed bracket in illegal position");
        }
    }

    private boolean lineContains(String wantToFind, String toCheck, int index1, int index2) {
        try {
            return wantToFind.equals(toCheck.substring(index1, index2));
        } catch (Exception e) {
        }
        ;
        return false;
    }

    /**
     * Checks if the String given is a number
     *
     * @param str the string to check
     * @return true if string is a number
     */
    public boolean isNumber(String str) {
        boolean result = false;
        try {
            Double.parseDouble(str);
            result = true;
        } catch (Exception e) {
        }
        return result;
    }

    /**
     * gets the number of errors
     *
     * @return the amount of errors
     */
    public int errors() {
        return errors.size();
    }

    public ArrayList<CompileError> getErrors() {
        return errors;
    }

    /**
     * lists the errors in the file
     */
    public void listErrors() {
        for (CompileError error : errors) {
            System.out.println(error);
        }
    }
}

File: js/compiler.js
var file = [];
var errors = [];

function start() {
    errors = [];
    file = document.getElementById("code").value.split('\n');
    document.getElementById("errors").value = "-------\n\nAdvanced-Planner Online Compiler\nBy John Elizarraras\n\n-------\n\nCompiling...Please Wait";
    compiles();
    setTimeout(showErrors, 500);
}

function addError(line, message) {
    errors.push("Error on line: " + line + " Error Message: " + message);
    errors.push("---")
}

/**
 * goes through the text field and checks each line
 */
function compiles() {
    for (var i = 0; i < file.length; i++) {
        var line = file[i];
        if (isCommand(i)) {
            switch (command(i).toLowerCase()) {
            case "run":
                break;
            case "loop":
                checkLoop(i);
                break;
            case "display":
                break;
            case "{":
                checkOpenBrackets(i);
                break;
            case "}":
                checkCloseBrackets(i);
                break;
            }
        }
    };
}

function showErrors() {
    var set = "-------\n\nAdvanced-Planner Online Compiler\nBy John Elizarraras\n\n-------\n"
    for (var index = 0; index < errors.length; index++) {
        var element = errors[index];
        set += "\n" + element;
    }
    if(errors.length == 0) set += "\nNo Errors\n---";
    set += "\nCompiled at " + Date().substring(16,25);
    document.getElementById("errors").value = set;
}

function lineContains(want, toCheck, p, q) {
    try {
        return want == toCheck.substring(p, q);
    } catch (err) {}
    return false;
}

function checkOpenBrackets(k) {
    try {
        if (!lineContains("loop", file[k - 1], 0, 4)) addError(k + 1, "Opend bracket without loop before");
    } catch (err) {
        addError(k + 1, "Open bracket without loop before");
    }
}

function checkCloseBrackets(k) {
    try {
        var needError = true;
        for (var i = k - 1; i >= 0; i--) {
            if (lineContains("loop", file[i], 0, 4)) {
                needError = false;
                break;
            } else if (lineContains("}", file[i], 0, 4)) addError(i + 1, "'}' before '{'");
        }
        if (needError) addError(index, "'}' with no opening bracket");
    } catch (err) {
        addError(k + 1, "Closed bracket in illegal position");
    }
}

function checkLoop(lineNum) {
    var mods = this.object(lineNum);
    if (mods.length < 2) addError(lineNum + 1, "Not Enough Modifiers");
    else if (mods.length > 2) addError(lineNum + 1, "Too Many Modifiers");
    else {
        if (Number.isNaN(mods[0].substring(0, mods[0].length - 1))) addError(lineNum + 1, "Time is not a number");
        switch (mods[0].substring(mods[0].length - 1)) {
        case "s":
            break;
        case "m":
            break;
        case "h":
            break;
        case "d":
            break;
        default:
            addError(lineNum + 1, "Not an acceptable time format");
            break;
        }
        if (isNaN(mods[1])) addError(lineNum + 1, "Second modifier is not a number");
        else {
            if (mods[1] < 0) addError(lineNum + 1, "Second modifier is negative");
        }
        try {
            if (!(file[lineNum + 1] == "{")) addError(lineNum + 2, "Is not a '{'");
            var needsError = true;
            for (var i = lineNum + 2; i < file.length; i++) {
                if (file[i] == "{") addError(i + 1, "Another '{' before '}'");
                else if (file[i] == "}") {
                    needsError = false;
                    break;
                }
            }
            if (needsError) addError(file.length + 1, "No Closing '}'");
        } catch (err) {
            addError(file.length + 1, "Loop doesn't open/close");
        }
    }
}

function object(line) {
    if (line < file.length && line > -1) {
        var a = [];
        var curLine = file[line];
        var startFound = false;
        var start = 0;
        for (var i = 0; i < curLine.length; i++) {
            if (curLine.substring(i, i + 1) === " ") {
                start = i;
                startFound = true;
                break;
            }
        }
        if (!startFound) return [];
        var str = curLine.substring(start + 1);
        var build = "";
        var lastCut = 0;
        for (var i = 0; i < str.length; i++) {
            var c = str.substring(i, i + 1);
            if (c === " ") {
                a.push(build.trim());
                build = "";
                lastCut = i;
            } else build += c.trim();
        }
        a.push(str.substring(lastCut).trim());
        return a;

    } else return [];
}

function isCommand(line) {
    var command = this.command(line);
    var res = true;
    switch (command.toLowerCase()) {
    case "loop":
        break;
    case "run":
        break;
    case "{":
        break;
    case "}":
        break;
    case "alias":
        break;
    case "display":
        break;
    default:
        addError(line + 1, "This is not a command");
        result = false;
        break;
    }
    return res;
}

function command(line) {
    var command = "";
    for (var i = 0; i < file[line].length; i++) {
        var str = file[line].substring(i, i + 1);
        if (!(str === " " || str === ".")) {
            command += str;
        } else {
            break;
        }
    };
    return command;
}

function pause() {
    return "";
}
